### PR TITLE
feat(#2259): Reveal the problem with getting several versions of the same object

### PR DESCRIPTION
**Pull request by volodya-lombrozo**
Reveal the problem with getting several versions of the same object. What have been done:

1. Refactoring of `DiscoveryMojo` and `DiscoveryMojoTest` classes.
2. Add test that reveals the problem with getting of several versions of the same object
3. Add one more puzzle to solve that problem

Related to #2259

### CODE REVIEW

**Comment by volodya-lombrozo** [comment]
@maxonfjvipon Could you have a look, please?

File: eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java [hunk 1]
```diff
@@ -55,17 +54,19 @@
...
-                this.scopedTojos().add(name).withDiscoveredAt(src);
-                discovered.add(name);
-            }
-            tojo.withDiscovered(names.size());
+            tojo.withDiscovered(
+                (int) this.discover(src)
+                    .stream()
+                    .filter(name -> !name.isEmpty())
+                    .peek(name -> this.scopedTojos().add(name).withDiscoveredAt(src))
+                    .peek(discovered::add)
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:67]: @volodya-lombrozo maybe we can join this line and line above into single one? Something like `.peek(name -> discovered.add(this.scopedTojos().add(name).withDiscoveredAt(src)))`

File: eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java [hunk 2]
```diff
@@ -55,17 +54,19 @@
...
         final Collection<String> discovered = new HashSet<>(1);
         for (final ForeignTojo tojo : tojos) {
             final Path src = tojo.optimized();
-            final Collection<String> names = this.discover(src);
-            for (final String name : names) {
-                this.scopedTojos().add(name).withDiscoveredAt(src);
-                discovered.add(name);
-            }
-            tojo.withDiscovered(names.size());
+            tojo.withDiscovered(
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:62]: @volodya-lombrozo what's the reason to turn these lines into stream pipeline? Is it about speed optimization or readability?

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]

File: eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java (hunk 1 above)
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:67]: @maxonfjvipon  `discovered.add(String)` accepts `String` object, `this.scopedTojos().add(name).withDiscoveredAt(src))` returns `ForeignTojo` object. So, it's impossible. Moreover, I don't see a problem in using `peek` several times.

File: eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java (hunk 2 above)
Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:62]: @maxonfjvipon It's is about code [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity):
> Some studies[[12]](https://en.wikipedia.org/wiki/Cyclomatic_complexity#cite_note-schroeder99-14) find a positive correlation between cyclomatic complexity and defects: functions and methods that have the highest complexity tend to also contain the most defects

Before it was 2 nested loops, now it's just a single loop

Comment (volodya-lombrozo) [reply eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:62]: @maxonfjvipon It's is about code [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity):
> Some studies[[12]](https://en.wikipedia.org/wiki/Cyclomatic_complexity#cite_note-schroeder99-14) find a positive correlation between cyclomatic complexity and defects: functions and methods that have the highest complexity tend to also contain the most defects

Before it was 2 nested loops, now it's just a single loop

Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java:62]: @volodya-lombrozo fair enough

**Review by maxonfjvipon — APPROVED** [review]

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**